### ISSUE TITLE
Issue with BIOMOD_ModelingOptions in biomod2 v4.2.6.2

### ISSUE THREAD

**Issue by letro81** [biomodhub/biomod2#650]
I am exploring biomod2 for SDM and I encountered unexpected behavior with the BIOMOD_ModelingOptions function in biomod2 versions 4.2.6.2 (which I currently have installed). When I call BIOMOD_ModelingOptions to define my model parameters, the function is returning an object of class "character" instead of the expected "BIOMOD.options" S4 object. This is preventing the subsequent BIOMOD_Modeling function from executing, as it requires bm.options to be a BIOMOD.options object.  When I downgraded the version to 4.2.4, the desired object class is coming but in the older versions, I am not able to use blockCV anbd advanced FR settings. 

My R session details are as follows: R version: 4.5.1 (2025-06-13 ucrt) Platform: x86_64-w64-mingw32/x64 biomod2 version: 4.2-6-2.

If somebody can kindly help me sesolve this, I will be immesnely grateful.

Thank you,

Letro

**Comment by letro81** [comment]
Hi Maya,

I am quite new to biomod2. Your earlier information was very helpful in looking for the right functions. 
I am working on species distribution at national scale and I felt I have a good presence distribution (over 500) and good set of variables (at least 11 after VIF). When I run simple automated approach, the results look good, except for Random Forest which seems to be overfiutting. 

I attempted to build robust species distribution models using `BIOMOD2` with user-defined pseudo-absences, k-fold cross-validation, and tuned model settings. Despite various attempts (see streamlined script below), the model run (`BIOMOD_Modeling`) consistently fails. Any suggestions or solutions would be much appreciated.

Your guidance and help will be of immense help. 


# --- 1. SETUP INPUT DATA ---

# Filter predictors
myExpl_final <- myExpl[[final_selected_vars]]

# Get presence coordinates
myRespXY <- as.data.frame(crds(presence_vect_proj))
colnames(myRespXY) <- c("x", "y")

# Generate pseudo-absence mask (exclude 2km buffer around presence points)
presence_buffer <- terra::buffer(presence_vect_proj, width = 2000)
buffer_raster <- rasterize(presence_buffer, master_mask, field = 1, background = 0)
pa_mask <- terra::ifel(buffer_raster == 1, NA, master_mask)

# Sample pseudo-absence points (3x number of presence)
set.seed(42)
pa_coords <- terra::spatSample(pa_mask,
                               size = 3 * nrow(myRespXY),
                               method = "random",
                               as.df = TRUE,
                               xy = TRUE,
                               na.rm = TRUE)[, c("x", "y")]

# Combine coordinates and define response
all_coords <- rbind(myRespXY, pa_coords)
myResp <- c(rep(1, nrow(myRespXY)), rep(0, nrow(pa_coords)))

# Extract environmental data
env_data_all <- terra::extract(myExpl_final, all_coords, xy = FALSE)

# Final modeling dataframe
modeling_df <- na.omit(data.frame(myResp, all_coords, env_data_all))
cat("Final data dimensions:", dim(modeling_df), "\n")
table(modeling_df$myResp)

# Extract cleaned inputs
myResp_clean <- modeling_df$myResp
myExpl_clean <- modeling_df[, names(myExpl_final)]
myRespXY_clean <- modeling_df[, c("x", "y")]

# --- 2. FORMAT DATA FOR BIOMOD2 ---

pa.user.table <- data.frame(PA1 = (myResp_clean == 0))

myBiomodData <- BIOMOD_FormatingData(
  resp.var = myResp_clean,
  expl.var = myExpl_clean,
  resp.xy = myRespXY_clean,
  resp.name = myRespName,
  PA.strategy = 'user.defined',
  PA.user.table = pa.user.table
)

# --- 3. DEFINE CROSS-VALIDATION TABLE ---

# Generate stratified 5-fold CV
folds <- createFolds(myBiomodData@data.species, k = 5, list = TRUE, returnTrain = FALSE)
myCVtable <- matrix(FALSE, nrow = length(myBiomodData@data.species), ncol = 5)
for (i in 1:5) myCVtable[folds[[i]], i] <- TRUE
colnames(myCVtable) <- paste0("_PA1_RUN", 1:5)

# --- 4. MODEL OPTIONS (TUNING ATTEMPT) ---

user.val <- list(
  MAXENT.binary.MAXENT.MAXENT = list(for_all_datasets = list(betamultiplier = 3)),
  RF.binary.randomForest.randomForest = list(for_all_datasets = list(ntree = 1000, mtry = 3, nodesize = 10)),
  GAM.binary.mgcv.gam = list(for_all_datasets = list(k = 6)),
  GLM.binary.stats.glm = list(for_all_datasets = list(type = "quadratic", interaction.level = 0)),
  GBM.binary.gbm.gbm = list(for_all_datasets = list(n.trees = 1000, interaction.depth = 3)),
  CTA.binary.rpart.rpart = list(for_all_datasets = list(method = "class"))
)

myBiomodOptions <- bm_ModelingOptions(
  data.type = "binary",
  models = c("MAXENT", "RF", "GAM", "GBM", "CTA", "GLM"),
  strategy = "user.defined",
  user.val = user.val
)

# --- 5. RUN BIOMOD MODELING ---

myBiomodModelOut <- BIOMOD_Modeling(
  bm.format = myBiomodData,
  modeling.id = "DholeModelingRun1",
  models = c("MAXENT", "RF", "GAM", "GBM", "CTA", "GLM"),
  OPT.user = myBiomodOptions,
  CV.strategy = "user.defined",
  CV.user.table = myCVtable,
  metric.eval = c("TSS", "ROC"),
  var.import = 3
)



My session is as follows;
> sessionInfo()
R version 4.4.3 (2025-02-28 ucrt)
Platform: x86_64-w64-mingw32/x64
Running under: Windows 11 x64 (build 26100)

Matrix products: default

locale:
[1] LC_COLLATE=English_Australia.utf8  LC_CTYPE=English_Australia.utf8   
[3] LC_MONETARY=English_Australia.utf8 LC_NUMERIC=C                      
[5] LC_TIME=English_Australia.utf8    

time zone: Australia/Sydney
tzcode source: internal

attached base packages:
[1] splines   stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] caret_7.0-1          lattice_0.22-7       ecospat_4.1.2        corrplot_0.95       
 [5] pROC_1.18.5          tidyterra_0.7.2      ggplot2_3.5.2        dplyr_1.1.4         
 [9] usdm_2.1-7           xgboost_1.7.11.1     randomForest_4.7-1.2 maxnet_0.1.4        
[13] earth_5.3.4          plotmo_3.6.4         plotrix_3.8-4        Formula_1.2-5       
[17] gbm_2.2.2            mgcv_1.9-1           nlme_3.1-167         gam_1.22-5          
[21] foreach_1.5.2        mda_0.5-5            class_7.3-23         rpart_4.1.24        
[25] nnet_7.3-20          biomod2_4.2-6-2      terra_1.8-60        


Thank you,

**Comment by letro81** [comment]
Hi Maya,
Thanks; 
The warning after myCVtable rung was as below;

myCVtable <- bm_CrossValidation(
+   bm.format = myBiomodData,
+   strategy = 'kfold',
+   k = 5,
+   nb.rep = 1,
+   strat = 'both' # <-- THIS IS THE FIX
+ )

Checking Cross-Validation arguments...
   > k-fold cross-validation selection
   !!! Some calibration dataset do not have both presences and absences:  _PA1_RUN1, _PA1_RUN2, _PA1_RUN3, _PA1_RUN4, _PA1_RUN5
   !!! Some validation dataset do not have both presences and absences:  _PA1_RUN1, _PA1_RUN2, _PA1_RUN3, _PA1_RUN4, _PA1_RUN5
Warning messages:
1: In bm_CrossValidation(bm.format = myBiomodData, strategy = "kfold",  :
  Some calibration repetion do not have both presences and absences
2: In bm_CrossValidation(bm.format = myBiomodData, strategy = "kfold",  :
  Some validation repetion do not have both presences and absences

and further more when I run the BIOMOD_Modeling, it says
 ***singfle value predicted 
! Note: [Name of model] failed!

That also leads to R session Abortion, and termination... Sam happens when I try again.

Thank you,

Letro

**Comment by MayaGueguen** [comment]
Huuum it seems that there is something going wrong in cross-validation selection 👀 

Could you share the output of : 
```
print(myBiomodData)
plot(myBiomodData)
summary(myBiomodData)
```
?

Maya

**Comment by letro81** [comment]
The outputs looks like this;


class(myBiomodData)
[1] "BIOMOD.formated.data.PA"
attr(,"package")
[1] "biomod2


print(myBiomodData

-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-= dhole Data Formating -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=


Checking Pseudo-absence selection arguments...

   > User defined pseudo absences selection

 ! Some NAs have been automatically removed from your data
      ! No data has been set aside for modeling evaluation
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-= Done -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
> print(myBiomodData)

-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-= BIOMOD.formated.data -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

dir.name =  .

sp.name =  dhole

	 504 presences,  1524 true absences and  0 undefined points in dataset


	 11 explanatory variables

     bio15          corez_dist         dist_fmu      composite_prey_richness
 Min.   :0.0415   Min.   :0.00000   Min.   :0.0000   Min.   :0.04522        
 1st Qu.:0.4094   1st Qu.:0.01250   1st Qu.:0.0921   1st Qu.:0.34212        
 Median :0.4991   Median :0.09156   Median :0.2172   Median :0.58384        
 Mean   :0.5091   Mean   :0.16824   Mean   :0.2614   Mean   :0.55423        
 3rd Qu.:0.5894   3rd Qu.:0.25666   3rd Qu.:0.3952   3rd Qu.:0.75695        
 Max.   :0.9726   Max.   :0.96387   Max.   :0.9697   Max.   :0.96101        
                                                                            
      bio3        settlement_dist       slope         livestock_density
 Min.   :0.1147   Min.   :0.00000   Min.   :0.04514   Min.   :0.00000  
 1st Qu.:0.4751   1st Qu.:0.01263   1st Qu.:0.31350   1st Qu.:0.02539  
 Median :0.5487   Median :0.04581   Median :0.38846   Median :0.03813  
 Mean   :0.5519   Mean   :0.09727   Mean   :0.38648   Mean   :0.07227  
 3rd Qu.:0.6268   3rd Qu.:0.12682   3rd Qu.:0.46087   3rd Qu.:0.09306  
 Max.   :0.9590   Max.   :0.92408   Max.   :0.88228   Max.   :1.00000  
                                                                       
     aspect          water_dist            landcover_composite
 Min.   :0.08122   Min.   :0.00000   NoData          :   0    
 1st Qu.:0.37385   1st Qu.:0.02477   Forests         :1460    
 Median :0.52423   Median :0.08079   Shrubland       : 235    
 Mean   :0.52397   Mean   :0.11082   Cultivated_land :  52    
 3rd Qu.:0.67228   3rd Qu.:0.16817   Alpine_scree    :  76    
 Max.   :0.95450   Max.   :0.81865   Barrenland      : 112    
                                     Snow_and_Glacier:  93    


 1 Pseudo Absences dataset available ( PA1 ) with  1020 (PA1) pseudo absences

plot(myBiomodData)

<SpatRaster> resampled to 500434 cells.
$data.vect
 class       : SpatVector 
 geometry    : points 
 dimensions  : 3552, 2  (geometries, attributes)
 extent      : 129319.5, 458319.5, 2956247, 3122813  (xmin, xmax, ymin, ymax)
 coord. ref. :  
 names       :  resp         dataset
 type        : <num>           <chr>
 values      :    10 Initial dataset
                  10 Initial dataset
                  10 Initial dataset

$data.label
                              9                              10 
                "**Presences**"       "Presences (calibration)" 
                             11                              12 
       "Presences (validation)"        "Presences (evaluation)" 
                             19                              20 
            "**True Absences**"   "True Absences (calibration)" 
                             21                              22 
   "True Absences (validation)"    "True Absences (evaluation)" 
                             29                              30 
          "**Pseudo-Absences**" "Pseudo-Absences (calibration)" 
                             31                               1 
 "Pseudo-Absences (validation)"                    "Background" 

$data.plot

summary(myBiomodData)

      dataset run   PA Presences True_Absences Pseudo_Absences Undefined
1     initial  NA <NA>       504          1524               0         0
2 calibration  NA  PA1

Thank you.

**Comment by MayaGueguen** [comment]
Hello Letro,

Glad to see you're making progress 💮 

As for your setting of modeling options, you were quite right about it 🙂 
You just need to change the `_allData_allRun` to `for_all_datasets` when building your list of new options.
```
## For example :
user.MAXENT_official <- list('for_all_datasets' = list(
  betamultiplier = 2.0, 
  linear = TRUE,        
  quadratic = TRUE,
  product = FALSE,
  threshold = FALSE,    
  hinge = FALSE
))
```

It is a particular keyword which biomod2 understands as : apply this set of parameters to all of my different datasets (_all PA and cross-validation combinations_).

👀 

Maya
